src/FakeItEasy/SelfInitializedFakes/MethodInfoConverter.cs

```diff
@@ -0,0 +1,118 @@
+                (string)reader.Value,
+                nameof(MethodInfoData.QualifiedTypename));
+            reader.Read();
+            var typeName = (string)reader.Value;
+            reader.Read();
+            Debug.Assert(
+                (string)reader.Value == nameof(MethodInfoData.MethodName),
+                "Unexpected Json property name {0}. Expected: {1}.",
+                (string)reader.Value,
+                nameof(MethodInfoData.MethodName));
+
+            reader.Read();
+            var methodName = (string)reader.Value;
+            reader.Read();
+            Debug.Assert(
+                (string)reader.Value == nameof(MethodInfoData.ParameterTypes),
+                "Unexpected Json property name {0}. Expected: {1}.",
+                (string)reader.Value,
+                nameof(MethodInfoData.ParameterTypes));
+            reader.Read();
+            var parameterTypeNames = ((string)reader.Value)?.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
+
+            // Need to advance the file read pointer past the end of current object
+            reader.Read(); // EndArray
+            reader.Read(); // EndObject
+
+            var type = Type.GetType(typeName);
+            var parameterTypes = parameterTypeNames == null ? new Type[0] : parameterTypeNames.Select(name => Type.GetType(name)).ToArray();
+            return type.GetMethod(methodName, parameterTypes);
+        }
+
+        /// <summary>
+        /// Writes the JSON representation of the object.
+        /// </summary>
+        /// <param name="writer">The <see cref="Newtonsoft.Json.JsonWriter"/> to write to.</param>
+        /// <param name="value">The value.</param>
+        /// <param name="serializer">The calling serializer.</param>
+        public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+        {
+            var methodInfo = (MethodInfo)value;
+            MethodInfoData data = new MethodInfoData
+            {
+                QualifiedTypename = methodInfo.DeclaringType.AssemblyQualifiedName,
+                MethodName = methodInfo.Name,
+                ParameterTypes = methodInfo.GetParameters().Select(p => p.ParameterType.AssemblyQualifiedName).ToArray()
+            };
+            writer.WriteStartObject();
+            writer.WritePropertyName(nameof(data.QualifiedTypename));
+            writer.WriteValue(data.QualifiedTypename);
+            writer.WritePropertyName(nameof(data.MethodName));
+            writer.WriteValue(data.MethodName);
+            writer.WritePropertyName(nameof(data.ParameterTypes));
+            writer.WriteStartArray();
+            for (int i = 0; i < data.ParameterTypes.Length; i++)
+            {
+                writer.WriteValue(data.ParameterTypes[i]);
+            }
+
+            writer.WriteEndArray();
+            writer.WriteEndObject();
+        }
+
+        private class MethodInfoData

```

Comment:
I think I was not clear. This class is a logicless data bag. On [line 78](https://github.com/FakeItEasy/FakeItEasy/pull/617/files/9ea652d2bc590f081f954c9974aff6171111ef74#diff-be7e0be8af3c2c37147cb11fa8358919R78) we instantiate it and fill it with data. Then we read the properties off the instance when writing the JSON. We could've just used the original data:

```c#
private const string QualifiedTypename = "QualifiedTypeName";
private const string MethodName = "MethodName";
private const string ParameterTypes = "ParameterTypes";

…

var methodInfo = (MethodInfo)value;
var qualifiedTypename = methodInfo.DeclaringType.AssemblyQualifiedName;
var methodName = methodInfo.Name;
var parameterTypes = methodInfo.GetParameters().Select(p => p.ParameterType.AssemblyQualifiedName).ToArray();
writer.WriteStartObject();
writer.WritePropertyName(QualifiedTypename);
writer.WriteValue(qualifiedTypename);
writer.WritePropertyName(MethodName);
writer.WriteValue(methodName);
writer.WritePropertyName(ParameterTypes);
writer.WriteStartArray();
```